File: 0113-path-sum-ii/0113-path-sum-ii.rb
# Definition for a binary tree node.
# class TreeNode
#   attr_accessor :val, :left, :right
#   def initialize(val = 0, left = nil, right = nil)
#     @val = val
#     @left = left
#     @right = right
#   end
# end
# @param {TreeNode} root
# @param {Integer} target_sum
# @return {Integer[][]}
def path_sum(root, target_sum)
  ans, running_sum = [], 0
  stack1 = root ? [root] : []
  stack2 = []
  curr = nil
  while stack1 != []
    curr = stack1.pop
    stack2 << curr
    running_sum += curr.val
    stack1 << curr.right if curr.right
    stack1 << curr.left if curr.left
    if !stack2[-1].left && !stack2[-1].right
      ans << stack2.map{ |node| node.val } if running_sum == target_sum
      while stack1 != [] && stack2 != [] && stack1[-1] != stack2[-1].right
        curr = stack2.pop
        running_sum -= curr.val
      end
    end
  end
  ans
end

File: 0113-path-sum-ii/0113-path-sum-ii.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *   val: number
 *   left: TreeNode | null
 *   right: TreeNode | null
 *   constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 *   }
 * }
 */
function pathSum(root: TreeNode | null, targetSum: number): number[][] {
  const ans: number[][] = []
  let runningSum = 0
  const stack1: TreeNode[] = root ? [root] : []
  const stack2: TreeNode[] = []
  let curr: TreeNode | null
  while (stack1.length) {
    curr = stack1.pop()
    stack2.push(curr)
    runningSum += curr.val
    if (curr.right) stack1.push(curr.right)
    if (curr.left) stack1.push(curr.left)
    if (!stack2[stack2.length - 1].left && !stack2[stack2.length - 1].right) {
      if (runningSum === targetSum) ans.push(stack2.map(node => node.val))
      while (stack1.length && stack2.length && stack1[stack1.length - 1] !== stack2[stack2.length - 1].right) {
        curr = stack2.pop()
        runningSum -= curr.val
      }
    }
  }
  return ans
};